src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+                throw new \RuntimeException('No sessions available.');
+            }
+
+            $session = $this->waitForNextAvailableSession();
+        }
+
+        return $this->database->session($session['name']);
+    }
+
+    /**
+     * Release a session back to the pool.
+     *
+     * @param Session $session The session.
+     */
+    public function release(Session $session)
+    {
+        $this->config['lock']->synchronize(function () use ($session) {
+            $item = $this->cacheItemPool->getItem($this->cacheKey);
+            $data = $item->get();
+
+            unset($data['inUse'][$session->name()]);
+            array_push($data['queue'], [
+                'name' => $session->name(),
+                'expiration' => $session->expiration()
+                    ?: $this->time() + SessionPoolInterface::SESSION_EXPIRATION_SECONDS
+            ]);
+            $this->cacheItemPool->save($item->set($data));
+        });
+    }
+
+    /**
+     * Clear the session pool.
+     */
+    public function clear()
+    {
+        $this->cacheItemPool->clear();
+    }
+
+    /**
+     * Set the database used to make calls to manage sessions.
+     *
+     * @param Database $database The database.
+     */
+    public function setDatabase(Database $database)
+    {
+        $this->database = $database;
+        $identity = $database->identity();
+        $this->cacheKey = sprintf(self::CACHE_KEY_TEMPLATE, $identity['instance'], $identity['database']);
+    }
+
+    /**
+     * Get the underlying cache implementation.
+     *
+     * @return CacheItemPoolInterface
+     */
+    public function cacheItemPool()
+    {
+        return $this->cacheItemPool;
+    }
+
+    /**
+     * Get the current unix timestamp.
+     *
+     * @return int
+     */
+    protected function time()
+    {
+        return time();
+    }
+
+    /**
+     * Builds out a list of timestamps indicating the start time of the intent
+     * to create a session.
+     *
+     * @param array $data
+     * @param bool $hasSession
+     * @return array
+     */
+    private function buildToCreateList(array $data, $hasSession)
+    {
+        $number = 0;
+        $toCreate = [];
+        $time = $this->time();
+        $count = $this->getSessionCount($data);
+
+        if ($count < $this->config['minSessions']) {
+            $number = $this->config['minSessions'] - $count;
+        } elseif (!$hasSession && !$data['queue'] && $count < $this->config['maxSessions']) {
+            $number++;
+        }
+
+        for ($i = 0; $i < $number; $i++) {
+            $toCreate[uniqid($time . '_')] = $time;
+        }
+
+        return $toCreate;
+    }
+
+    /**
+     * Purge any items in the to create queue that have been inactive for 20
+     * minutes or more.
+     *
+     * @param array $data
+     */
+    private function purgeOrphanedToCreateItems(array &$data)
+    {
+        foreach ($data['toCreate'] as $key => $timestamp) {
```

Comment:
This is here to handle cases where the process might stop unexpectedly while `acquire` is being called. If items are in the `toCreate` list for 20 minutes, it is almost certain it is due to the fact the process died (in most cases, this would not happen). Additionally, when we go to put the created items into the queue we also decrement the `toCreate` list by the count of sessions created in that process - so there should not be cases where we add new sessions to the queue while also not adjusting the `toCreate` count.

I do agree that if each process only creates at most one session that would help simplify things. A `warmPool` method sounds good to me.